### ISSUE TITLE
Boolean condition in if statement not recognized if saved in a variable

### ISSUE THREAD

**Issue by LefterisJP** [python/mypy#7586]
**Python**: 3.7
**Mypy**: 0.730

To easily see the problem please check the following example:

```python
from typing import Optional


def maybe_create() -> Optional[int]:
    return 5


def use(a: int) -> None:
    print(a)

# Create a, which can also be none here
a = maybe_create()
# We have a validity check but instead of directly adding it on the if statement
# it's first put in a boolean variable
valid = a is not None and a > 0
if valid:
    use(a)
```

This will return a mypy error `Argument 1 to "use" has incompatible type "Optional[int]"; expected "int"`

Even though there is a check right above the `use()` function, it seems to not be recognized just because the result of the check is saved in a variable.

Instead if the check is directly in the if statement there are no errors reported by mypy:

```python
from typing import Optional


def maybe_create() -> Optional[int]:
    return 5


def use(a: int) -> None:
    print(a)

# Create a, which can also be none here
a = maybe_create()
# Check if it's none directly in the if statement
if a is not None and a > 0:
    use(a)

```

Why can't mypy figure this out if another temporary variable is added in the mix?

You may ask why would anyone want to do this? One answer is readability:

```python
if (
    a_very_big_variable_name > another_very_big_variable name and
    (
        more_bigvariable_names != a_big constant or
        other_stuff >= 100
    )):
```

looks really ugly and is hard to read. If instead parts of the condition are saved in separate variables you can make the if check very nicely readable. But then it seems that mypy gets confused.

Is there anything we can do here to make this work better?

**Comment by gvanrossum** [comment]
This is just how it works. Otherwise mypy would have to remember the "meaning" of every value ever assigned to a variable, in case it's later used in a context where it would matter for type checking. And it would have to keep track of whether anything could change that meaning. For example this:
```py
valid = a is not None and a > 0
a = some_other_function(a)
if valid:
    use(a)
```
I propose to close this as won't fix.

**Comment by LefterisJP** [comment]
I had guessed that this is a reason but hoped it would be possible to track variables, at least to see if they get touched again or not. I don't know the implementation details of mypy so can not judge, which is I why I opened the issue to ask you guys.

**Comment by gvanrossum** [comment]
No, that is not easy.
